.yarn/patches/@metamask-assets-controllers-npm-38.2.0-40af2afaa7.patch

```diff
@@ -0,0 +1,36 @@
+diff --git a/dist/assetsUtil.cjs b/dist/assetsUtil.cjs
+index e90a1b6767bc8ac54b7a4d580035cf5db6861dca..2662df7869e9079f9379c4dd1ded40b5af4e71f2 100644
+--- a/dist/assetsUtil.cjs
++++ b/dist/assetsUtil.cjs
+@@ -3,6 +3,7 @@ var __importDefault = (this && this.__importDefault) || function (mod) {
+     return (mod && mod.__esModule) ? mod : { "default": mod };
+ };
+ Object.defineProperty(exports, "__esModule", { value: true });
++function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { newObj[key] = obj[key]; } } } newObj.default = obj; return newObj; } }
+ exports.fetchTokenContractExchangeRates = exports.reduceInBatchesSerially = exports.divideIntoBatches = exports.ethersBigNumberToBN = exports.addUrlProtocolPrefix = exports.getFormattedIpfsUrl = exports.getIpfsCIDv1AndPath = exports.removeIpfsProtocolPrefix = exports.isTokenListSupportedForNetwork = exports.isTokenDetectionSupportedForNetwork = exports.SupportedTokenDetectionNetworks = exports.formatIconUrlWithProxy = exports.formatAggregatorNames = exports.hasNewCollectionFields = exports.compareNftMetadata = exports.TOKEN_PRICES_BATCH_SIZE = void 0;
+ const controller_utils_1 = require("@metamask/controller-utils");
+ const utils_1 = require("@metamask/utils");
+@@ -221,7 +222,7 @@ async function getIpfsCIDv1AndPath(ipfsUrl) {
+     const index = url.indexOf('/');
+     const cid = index !== -1 ? url.substring(0, index) : url;
+     const path = index !== -1 ? url.substring(index) : undefined;
+-    const { CID } = await import("multiformats");
++    const { CID } = await Promise.resolve().then(() => _interopRequireWildcard(require("multiformats")));
+     // We want to ensure that the CID is v1 (https://docs.ipfs.io/concepts/content-addressing/#identifier-formats)
+     // because most cid v0s appear to be incompatible with IPFS subdomains
+     return {
+diff --git a/dist/token-prices-service/codefi-v2.mjs b/dist/token-prices-service/codefi-v2.mjs
+index e7eaad2cfa8b233c4fd42a51f745233a1cc5c387..416dab35e83334f0cafe23c9bd34d0a0380145b2 100644
+--- a/dist/token-prices-service/codefi-v2.mjs
++++ b/dist/token-prices-service/codefi-v2.mjs
+@@ -12,8 +12,8 @@ var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (
+ var _CodefiTokenPricesServiceV2_tokenPricePolicy;
+ import { handleFetch } from "@metamask/controller-utils";
+ import { hexToNumber } from "@metamask/utils";
+-import $cockatiel from "cockatiel";
+-const { circuitBreaker, ConsecutiveBreaker, ExponentialBackoff, handleAll, retry, wrap, CircuitState } = $cockatiel;
++
++import { circuitBreaker, ConsecutiveBreaker, ExponentialBackoff, handleAll, retry, wrap, CircuitState } from "cockatiel";
```

Comment:
The fix makes sense, but I don't understand why we're getting a default import here in the first place. 

@Mrtenz Is it possible that this is an artifact from `ts-bridge` processing mixed type vs. value imports in a single import statement?

This is the original code:
```ts
import {
  circuitBreaker,
  ConsecutiveBreaker,
  ExponentialBackoff,
  handleAll,
  type IPolicy,
  retry,
  wrap,
  CircuitState,
} from 'cockatiel';
```
> https://github.com/MetaMask/core/blob/main/packages/assets-controllers/src/token-prices-service/codefi-v2.ts#L4-L13

Reply:
This is not related to type imports, those are supported by `ts-bridge`.  I suspect this is caused by a difference in how Node.js and Webpack handle module detection... 😞

`ts-bridge` detects whether a package is CommonJS or ESM following Node.js' module resolution algorithm. If the package is detected as CommonJS, it checks what exports it has (using `cjs-module-lexer`, used by Node.js as well). If some CommonJS exports are not detected, the import is rewritten to a default import, as this is supposed to increase compatibility.

`cockatiel` has an ESM version, but it's only specified in the `module` field, which is not used by Node.js (but is used by Webpack). The CommonJS export seems to use a syntax that is not supported by `cjs-module-lexer` (though I need to verify this), so the exports are not detected.

There are several options here:

1. We apply this patch.
   - We could do this as temporary solution to get this PR merged either way.
3. We disable the `module` field in Webpack's module resolution.
   - This aligns it more with Node.js module resolution, but _may_ result in a larger bundle if many dependencies use the `module` field, but not `package.json` exports.
4. We submit a PR to `cockatiel` to use the `exports` field in `package.json` to allow Node.js to use the ESM version as well.
   - I will do this regardless, but I'm not sure how long it will take for this to be merged and released.

Reply:
Yet another CJS/ESM interop scenario to keep an eye out for. :( Really appreciate the breakdown and your looking into this. I had a couple of questions:

> If some CommonJS exports are not detected, the import is rewritten to a default import, as this is supposed to increase compatibility.

According to @sahar-fehri, this causes the following error:

> I had this [error](https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/103564/workflows/4db981ed-612d-416c-b53e-d0f87e5b291d/jobs/3860282) when doing `yarn build:storybook;`
```
export 'default' (imported as '$cockatiel') was not found in 'cockatiel'
```
Is storybook an outlier here, or will we always see this error if there is no default export?

> The CommonJS export seems to use a syntax that is not supported by `cjs-module-lexer`

Also, for this issue, is an upstream update pretty much the only possible fix?

Reply:
I guess you could say Webpack is the outlier because it looks at the `module` field. This is completely valid in Node.js for example:

```js
import $cockatiel from 'cockatiel';
const { circuitBreaker } = $cockatiel;

console.log(circuitBreaker); // [Function: circuitBreaker]
```

Because Node.js loads the CommonJS version, but Webpack loads the ESM version.

> Also, for this issue, is an upstream update pretty much the only possible fix?

I'm leaning towards configuring Webpack to stop using `module` for better Node.js compatibility. But for a proper fix where both are able to use ESM, without the need for a patch, it will need to be fixed upstream.

Reply:
We could apply this as patch temporarily until it's updated, rather than patching `assets-controller`:

- connor4312/cockatiel#102

Reply:
Really like the idea of patching closer to the root cause. 

@sahar-fehri To clarify, we'd want to remove this section of the `assets-controllers` patch altogether (lines 22-36), and instead add a new patch file for `cockatiel` that contains the changes in the cockatiel PR #102 linked above.

The Webpack issue is really... not great. I also think configuring Webpack would be the preferable solution here. Fixing or patching every upstream package that triggers this issue doesn't seem advisable or feasible.

Reply:
Thank you guys for looking into this 🙏  🙏 
Just to clarify; so its okay to keep the patch for cockatiel and merge it just temporarily; until this [PR](https://github.com/connor4312/cockatiel/pull/102) is merged?

Reply:
I just realised we need the `assets-controller` patch regardless, because it won't with the ESM version of `cockatiel` otherwise anyway. The `cockatiel` patch would need to be applied to the core repo, requiring a new release, etc.

So I think the easiest way forward for now is to keep this patch as it is, and hope that `cockatiel` gets updated quickly 😅.

Reply:
Sounds like a plan! Technically, anything downstream of `assets-controllers` would need to have this patch applied, but that's just our two clients so doesn't seem too terrible. In any case, it makes sense to avoid a controller release containing a patch. Hopefully the `cockatiel` maintainers will be responsive.